core/src/qos_service.rs

```diff
@@ -0,0 +1,239 @@
+//! Producer quality of service
+//!
+//!
+use solana_measure::measure::Measure;
+use solana_runtime::{
+    bank::Bank,
+    cost_model::{CostModel, TransactionCost},
+    cost_tracker::CostTrackerError,
+};
+use solana_sdk::transaction::{self, SanitizedTransaction, TransactionError};
+use std::sync::{Arc, RwLock};
+
+#[derive(Default)]
+pub struct QosServiceStats {
+    compute_cost_time: u64,
+    cost_tracking_time: u64,
+    selected_txs_count: u64,
+    retried_txs_per_block_limit_count: u64,
+    retried_txs_per_account_limit_count: u64,
+}
+
+impl QosServiceStats {
+    pub fn report(&mut self) {
+        datapoint_info!(
+            "qos-service-stats",
+            ("compute_cost_time", self.compute_cost_time, i64),
+            ("cost_tracking_time", self.cost_tracking_time, i64),
+            ("selected_txs_count", self.selected_txs_count, i64),
+            (
+                "retried_txs_per_block_limit_count",
+                self.retried_txs_per_block_limit_count,
+                i64
+            ),
+            (
+                "retried_txs_per_account_limit_count",
+                self.retried_txs_per_account_limit_count,
+                i64
+            ),
+        );
+    }
+}
+
+pub struct QosService {
+    cost_model: Arc<RwLock<CostModel>>,
+}
+
+impl QosService {
+    pub fn new(cost_model: Arc<RwLock<CostModel>>) -> Self {
+        Self { cost_model }
+    }
+
+    pub fn report_stats(&self) {}
```

Comment:
Is this meant to be empty?

Reply:
good eye! There is a follow up PR (#21191) to update stats reporting. Didn't want to smash into this one, hopefully to keep PRs clean. Removed this line.